core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1296,29 +1296,59 @@ class KafkaApis(val requestChannel: RequestChannel,
 
   def handleInitPidRequest(request: RequestChannel.Request): Unit = {
     val initPidRequest = request.body[InitPidRequest]
+    val transactionalId = initPidRequest.transactionalId
+
     // Send response callback
     def sendResponseCallback(result: InitPidResult): Unit = {
       val responseBody: InitPidResponse = new InitPidResponse(result.error, result.pid, result.epoch)
-      trace(s"Generated new PID ${result.pid} from InitPidRequest from client ${request.header.clientId}")
+      trace(s"Completed $transactionalId's InitPidRequest with result $result from client ${request.header.clientId}")
       requestChannel.sendResponse(new RequestChannel.Response(request, responseBody))
     }
-    txnCoordinator.handleInitPid(initPidRequest.transactionalId, initPidRequest.transactionTimeoutMs, sendResponseCallback)
-  }
 
-  def handleEndTransactionRequest(request: RequestChannel.Request): Unit = {
-    throw new UnsupportedOperationException
+    txnCoordinator.handleInitPid(transactionalId, initPidRequest.transactionTimeoutMs, sendResponseCallback)
   }
 
-  def handleAbortTransactionRequest(request: RequestChannel.Request): Unit = {
+  def handleEndTransactionRequest(request: RequestChannel.Request): Unit = {
     throw new UnsupportedOperationException
   }
 
   def handleAddPartitionToTransactionRequest(request: RequestChannel.Request): Unit = {
-    throw new UnsupportedOperationException
+    val addPartitionsToTxnRequest = request.body[AddPartitionsToTxnRequest]
+    val transactionalId = addPartitionsToTxnRequest.transactionalId
+    val partitionsToAdd = addPartitionsToTxnRequest.partitions
+
+    // Send response callback
+    def sendResponseCallback(error: Errors): Unit = {
+      val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(error)
+      trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: $error from client ${request.header.clientId}")
+      requestChannel.sendResponse(new RequestChannel.Response(request, responseBody))
+    }
+
+    txnCoordinator.handleAddPartitionsToTransaction(transactionalId,
+      addPartitionsToTxnRequest.pid,
+      addPartitionsToTxnRequest.epoch,
+      partitionsToAdd.asScala.toSet,
+      sendResponseCallback)
   }
 
   def handleAddOffsetsToTransactionRequest(request: RequestChannel.Request): Unit = {
```

Comment:
Mentioned before, but it seems that this should be handled by the consumer group coordinator per our design document: https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#heading=h.5695qbm2hne